src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs

```diff
@@ -251,11 +255,33 @@ private async ValueTask DisposeAsyncCore()
             }
             finally
             {
-                CloseStreamFromDispose(disposing: true);
+                await CloseStreamFromDisposeAsync().ConfigureAwait(false);
             }
             GC.SuppressFinalize(this);
         }
 
+        private async ValueTask CloseStreamFromDisposeAsync()
+        {
+            // Dispose of our resources if this StreamWriter is closable.
+            if (_closable && !_disposed)
+            {
+                try
+                {
+                    // Attempt to close the stream even if there was an IO error from Flushing.
+                    // Note that Stream.DisposeAsync() can potentially throw here (may or may not be
+                    // due to the same Flush error). In this case, we still need to ensure
+                    // cleaning up internal resources, hence the finally block.
+                    await _stream.DisposeAsync().ConfigureAwait(false);
+                }
+                finally
+                {
+                    _disposed = true;
+                    _charLen = 0;
+                    await base.DisposeAsync().ConfigureAwait(false);
```

Comment:
I stared at this for a long time and seems to me that this is the right thing to do, as per [the docs](https://learn.microsoft.com/en-us/dotnet/api/system.iasyncdisposable.disposeasync?view=net-7.0#remarks):
> An object must also call the DisposeAsync method of its base class if the base class implements IAsyncDisposable.

I don't know why we weren't doing this before. @stephentoub thoughts?

Reply:
What you have here looks good to me.